libraries/chain/platform_timer_asio_fallback.cpp

```diff
@@ -78,25 +80,30 @@ void platform_timer::start(fc::time_point tp) {
 }
 
 void platform_timer::expire_now() {
-   state_t expected = state_t::running;
-   if (_state.compare_exchange_strong(expected, state_t::timed_out)) {
+   timer_state_t expected{.state = state_t::running, .callback_in_flight = false};
+   if (_state.compare_exchange_strong(expected, timer_state_t{state_t::timed_out, true})) {
       call_expiration_callback();
+      _state.store(timer_state_t{state_t::timed_out, false});
    }
 }
 
 void platform_timer::interrupt_timer() {
-   state_t expected = state_t::running;
-   if (_state.compare_exchange_strong(expected, state_t::interrupted)) {
+   timer_state_t expected{.state = state_t::running, .callback_in_flight = false};
+   if (_state.compare_exchange_strong(expected, timer_state_t{state_t::interrupted, true})) {
       call_expiration_callback();
+      _state.store(timer_state_t{state_t::interrupted, false});
    }
 }
 
 void platform_timer::stop() {
-   const state_t prior_state = _state;
-   if(prior_state == state_t::stopped)
+   while (_state.load().callback_in_flight)
+      boost::core::sp_thread_pause();
```

Comment:
I'm worried this might be too early. Is it possible for a thread to `interrupt_timer()` but then call `stop()` before the CAS in `interrupt_timer()` occurs? If so it seems like we might have the same problem still? Because `stop()` would skip the the busy loop since `false` but then the CAS is entered which starts the callback.

So maybe we need another CAS here in `stop() `prior to busy waiting on `callback_in_flight`. The CAS purpose is to atomicly prevent `expire_now()` & `interrupt_timer()` from entering the `call_expiration_callback()`

Reply:
Very good point. Thanks.

Reply:
https://github.com/AntelopeIO/spring/pull/1606/commits/c509c731ae023543c79cffb710a7df9281b130dd